feat(aft_netif): Introduction of esp_netif with FreeRTOS+TCP stack

david-cermak opened the pull request:
**AFpT**: **A**mazon **F**reeRTOS plus **T**CP: network stack now supports multiple network interfaces with multiple endpoints!

## Related

* FreeRTOS+TCP ~~changes" https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/commit/d12c0d5c2e3fcc851eb2099fa10cf44bcfe7257b~~ (used upstream [v4.4](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/releases/tag/V4.4.0) )

## Task list

- [x] Support via `esp_netif`
- [x] Make flexible data path common for wifi and Ethernet reusing IDF glues
- [x] Support FreeRTOS SMP (added notes about thread safety)

## Next PRs

- [ ] Support better Network Buffer management

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **High Risk**
> New full networking stack and custom esp_netif implementation on the critical WiFi/DHCP/IP path; many APIs return NOT_SUPPORTED and SMP/critical-section shims affect stack-wide locking behavior.
> 
> **Overview**
> Adds a new **`freertos_tcp`** component that vendors **FreeRTOS-Plus-TCP** (submodule) and wires it into ESP-IDF as an **alternative TCP/IP stack** to lwIP for netifs created through this integration.
> 
> **`esp_netif` custom implementation** (`ESP_NETIF_PROVIDE_CUSTOM_IMPLEMENTATION` via Kconfig) owns lifecycle, DHCP client control, link up/down, and **IP_EVENT** got/lost IP posting. It uses **FreeRTOS_IPInit_Multi()**, a shared WiFi/Ethernet netstack glue (`interface.c`), and **`esp_netif_tcpip_exec()`** marshalled onto the IP task. The **port** layer provides **`NetworkInterface.c`** (copy-based RX into stack buffers, TX via `esp_netif_transmit`), **`FreeRTOSIPConfig.h`** (IPv4 default, SMP spinlocks), and app hooks (RNG, hostname).
> 
> Repo plumbing: **`aft` CI** build for `test_app` on esp32/esp32c3/esp32s3, commit scope **`aft_netif`**, copyright rules for `port/`. **v0.1.0** scope is **IPv4, single STA-style interface**; IPv6/DHCP server/multi-netif are largely stubbed or unsupported in `esp_netif_impl.c`.
> 
> Includes a **simple TCP client example**, **Unity + pytest** test app (TCP/UDP/DNS/DHCP), and README documenting architecture and limitations.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2a8f84f0939ce7c42e7d16fdf08547167bfb8604. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

david-cermak committed:
feat(aft_netif): Introduction of esp_netif with FreeRTOS+TCP stack

euripedesrocha commented on components/freertos_tcp/examples/simple/main/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+#set(fdir /home/david/esp/idf/components/freertos/FreeRTOS-Kernel/include/freertos)
```
Clean up.

euripedesrocha reviewed (commented):
We need to add documentation for this and make the limitations clear for this initial version.

david-cermak commented:
Cleaned-up, added README, documented basic ideas and limitations, thread safety and added simple tests (tcp, udp, dns), build only check in CI.

I think that this could be published as an MVP -- ready for review.

euripedesrocha commented on components/freertos_tcp/esp_netif/esp_netif_impl.c:
```diff
@@ -0,0 +1,577 @@
+#include "esp_event.h"
+#include "esp_idf_version.h"
+
+#define MAX_ENDPOINTS_PER_NETIF     1  // MVP: single interface, single endpoint
```
Should this be configurable?

david-cermak replied:
I'd keep it at `1` at least for this initial version. The thing is that LWIP has always 1 enpoint per netif, so we're kinda use to this setup.

euripedesrocha commented on components/freertos_tcp/esp_netif/esp_netif_impl.c:
```diff
@@ -0,0 +1,577 @@
+    }
+    netif->config.init_fn = cfg->init_fn;
+    if (!cfg->init_fn(netif_count++, &netif->aft_netif)) {
+        free(netif);
```
Should log error before return?

david-cermak replied:
added error log

euripedesrocha commented on components/freertos_tcp/README.md:
```diff
@@ -0,0 +1,278 @@
+
+## MVP Status
+
+**This is a Minimum Viable Product (MVP) release focusing on IPv4-only, single network interface functionality.**
```
Not sure about the MVP terminology adoption here. We can point the minimal, and beta stage of the component by the versioning semantics perhaps?

david-cermak replied:
Replaced MVP with something like `~initial version`

euripedesrocha commented on components/freertos_tcp/README.md:
```diff
@@ -0,0 +1,278 @@
+The component consists of three main layers:
+
+### 1. FreeRTOS-Plus-TCP Core (`FreeRTOS-Plus-TCP/`)
+- Upstream library (Git submodule)
```
Which version are we integrating?

david-cermak replied:
Updated -> `v4.4` https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/releases/tag/V4.4.0

euripedesrocha commented on components/freertos_tcp/README.md:
````diff
@@ -0,0 +1,278 @@
+idf.py build flash monitor
+```
+
+## File Structure
````
Is this necessary?

david-cermak replied:
Removed

euripedesrocha commented on components/freertos_tcp/README.md:
```diff
@@ -0,0 +1,278 @@
+| `AFPT_IPV4` | Y | Enable IPv4 stack (required for MVP) |
+| `AFPT_IPV6` | **N** | Enable IPv6 stack (experimental, not in MVP) |
+
+### Key Stack Parameters (in `FreeRTOSIPConfig.h`)
```
Am I correct that at this point we still need to keep LWIP? Or it is possible to disable LWIP and use only the FreeRTOS+TCP stack?

david-cermak replied:
LWIP is a core component, so by default it's still there, but not routed via `esp-netif` (still usable directly).
LWIP can also be disabled (independently), but I didn't add any Kconfig dependency to disable it at this point.

The thing is that it'll be stripped by the linker if the app doesn't use it.

euripedesrocha replied:
I'm aware about stripping. My concern here was more to clarify to users the scenario. Not sure to which extent we need to go though.

david-cermak replied:
Good point, have added a section to the readme, explaining how it works with LWIP

euripedesrocha commented on components/freertos_tcp/test_app/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# CMakeLists in this exact order for cmake to work correctly
+cmake_minimum_required(VERSION 3.5)
+
+set(EXTRA_COMPONENT_DIRS
```
Concentrate the dependencies in the yml for component manager is a cleaner choid IMO.

david-cermak replied:
Removed

david-cermak commented:
Thanks for the review @euripedesrocha ! Comments have been addressed, PTAL again.
Also switched to the latest upstream release `v4.4`, which changed API (and internals), so had to update some port layers.

euripedesrocha commented on components/freertos_tcp/test_app/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# The following lines of boilerplate have to be in your project's
+# CMakeLists in this exact order for cmake to work correctly
+cmake_minimum_required(VERSION 3.5)
```
```suggestion
cmake_minimum_required(VERSION 3.16)
```

If not mistaken cmake will give warning about that version

david-cermak replied:
that's correct. this component is only for IDF >= v5.5
